plugins/inputs/opensearch_query/opensearch_query.go

```diff
@@ -0,0 +1,327 @@
+
+	for _, agg := range o.Aggregations {
+		wg.Add(1)
+		go func(agg osAggregation) {
+			defer wg.Done()
+			err := o.osAggregationQuery(acc, agg)
+			if err != nil {
+				acc.AddError(fmt.Errorf("opensearch query aggregation %s: %s ", agg.MeasurementName, err.Error()))
+			}
+		}(agg)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (o *OpensearchQuery) osAggregationQuery(acc telegraf.Accumulator, aggregation osAggregation) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(o.Timeout))
+	defer cancel()
+
+	searchResult, err := o.runAggregationQuery(ctx, aggregation)
+	if err != nil {
+		return err
+	}
+
+	return searchResult.GetMetrics(acc, aggregation.MeasurementName)
+}
+
+func init() {
+	inputs.Add("opensearch_query", func() telegraf.Input {
+		return &OpensearchQuery{
+			Timeout:             config.Duration(time.Second * 5),
+			HealthCheckInterval: config.Duration(time.Second * 10),
+		}
+	})
+}
+
+func (o *OpensearchQuery) runAggregationQuery(ctx context.Context, aggregation osAggregation) (*AggregationResponse, error) {
+	now := time.Now().UTC()
+	from := now.Add(time.Duration(-aggregation.QueryPeriod))
+	filterQuery := aggregation.FilterQuery
+	if filterQuery == "" {
+		filterQuery = "*"
+	}
+
+	aq := &Query{
+		Size:         0,
+		Aggregations: aggregation.aggregation,
+		Query:        nil,
+	}
+
+	boolQuery := &BoolQuery{
+		FilterQueryString: filterQuery,
+		TimestampField:    aggregation.DateField,
+		TimeRangeFrom:     from,
+		TimeRangeTo:       now,
+		DateFieldFormat:   aggregation.DateFieldFormat,
+	}
+
+	aq.Query = boolQuery
+	req, err := json.Marshal(aq)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %s", err)
+	}
+	o.Log.Debugf("{\"body\": %s}", string(req))
+
+	searchRequest := &opensearchapi.SearchRequest{
+		Body:    strings.NewReader(string(req)),
+		Index:   []string{aggregation.Index},
+		Timeout: time.Duration(o.Timeout),
+	}
+
+	resp, err := searchRequest.Do(ctx, o.osClient)
+	if err != nil {
+		return nil, err
+	}
+	if resp.IsError() {
+		return nil, fmt.Errorf("Opensearch SearchRequest failure: [%d] %s", resp.StatusCode, resp.Status())
+	}
+	defer resp.Body.Close()
+
+	var searchResult AggregationResponse
+
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&searchResult)
+	if err != nil {
+		return nil, err
+	}
+
+	return &searchResult, nil
+}
+
+// getMetricFields function returns a map of fields and field types on Elasticsearch that matches field.MetricFields
+func (o *OpensearchQuery) getMetricFields(ctx context.Context, aggregation osAggregation) (map[string]string, error) {
+	mapMetricFields := make(map[string]string)
+	fieldMappingRequest := opensearchapi.IndicesGetFieldMappingRequest{
+		Index:  []string{aggregation.Index},
+		Fields: aggregation.MetricFields,
+	}
+
+	response, err := fieldMappingRequest.Do(ctx, o.osClient)
+	if err != nil {
+		return nil, err
+	}
+
+	// Bad request; move on
+	if response.StatusCode == 400 {
```

Comment:
Yeah, it works fine with checking if `status != 200`.  There are other cases where it might be a 201 or 202, but I don't think they apply here.  Good catch.